File: Data Structures & Algorithms/Backtracking/template.py
def is_valid_state(state):
  # check is it is a valid solution
  
  return True

def get_candidates(state):
  return []

# recursive function that calls the earlier one and checks the search
# makes a deep copy
def search(state, solutions):
  if is_valid_state(state):
    solutions.append(state.copy())
    # return

  for candidates in get_candidates(state):
    state.add(candidate)
    search(state, solutions)
    state.remove(candidate)

# this is what the leetcode problem want us to write 
def solve():
  solutions = []
  state = set()
  search(state, solutions)
  return solutions
  

File: Data Structures & Algorithms/Backtracking/template.cpp
#include <vector>
#include <unordered_set>

using State = std::unordered_set<int>;
using Candidate = int;

bool isValid(const State& state) {
  return true;
}

std::vector<Candidate> getCandidates(const State& state) {
  return {};
}

viod search(State& state, std::vector<State>& solutions) {
  if (isValidState(state)) {
        // In C++, pushing a state into a vector automatically creates a deep copy
        // of that state. This is the equivalent of `solutions.append(state.copy())`.
        solutions.push_back(state);
        // return; // Uncomment if you want to stop searching deeper once a solution is found
    }

    for (const auto& candidate : getCandidates(state)) {
        state.insert(candidate);    // Python: state.add(candidate)
        search(state, solutions);
        state.erase(candidate);     // Python: state.remove(candidate)
    }
}

std::vector<State> solve(){
  std::vector<State> solutions;
  State state;
  search(state, solutions);
  return solutions;
}
